Update wording for layout_left_padded (#409)
ORNL/cpp-proposals-pub: layout_padded/layout_padded.md

  * Simplify `layout_{left,right}_padded::mapping` constructor conditions

>   * Add default constructors for `layout_{left,right}_padded::mapping`; `=default` for _`static-padding-stride`_` != dynamic_extent`, not `=default` otherwise

Reviewer: If this was a new change, should we consider adding it instead to the R4 list (which doesn't exist yet, but will)?

[8]{.pnum} *Preconditions:*
If `extents_type::rank()` is greater than one
and `padding_value` does not equal `dynamic_extent`,
then the least multiple of `padding_stride`
greater than to equal to `ext.extent(0)`
is representable as a value of type `index_type`.

[9]{.pnum} *Effects:*

  * [9.1]{.pnum} Direct-non-list-initializes _`extents_`_ with `ext`, and

>   * [9.2]{.pnum} if _`static-padding-stride`_ is not equal to `dynamic_extent` direct-non-list-initialized _`stride-1`_ with `ext.extent(0)`.

Reviewer: ```suggestion
  * [9.2]{.pnum} if _`static-padding-stride`_ is not equal to `dynamic_extent`,
      direct-non-list-initializes _`stride-1`_ with `ext.extent(0)`.
```

```c++
template<class OtherIndexType>
constexpr mapping(const extents_type& ext, OtherIndexType pad);
```

[10]{.pnum} *Constraints:*

* [10.1]{.pnum} `is_convertible_v<OtherIndexType, index_type>` is `true`.

* [10.2]{.pnum} `is_nothrow_constructible_v<index_type, OtherIndexType>` is `true`.

[11]{.pnum} *Preconditions:*

  * [11.1]{.pnum} `pad` is representable as a value of type `index_type`,

>   * [11.2]{.pnum} `extents_type::`_`index-cast`_`(pad)` is greater than zero,

Reviewer: Resolved; leave in place.

  * [11.3]{.pnum} if `extents_type::rank()` is greater than one,
      then the least multiple of `pad`
      greater than or equal to `ext.extent(0)`
      is representable as a value of type `index_type`, and

  * [11.4]{.pnum} if `padding_value` is not equal to `dynamic_extent`, `padding_value` equals `extents_type::`_`index-cast`_`(pad)`.

[12]{.pnum} *Effects:*

  * [9.1]{.pnum} Direct-non-list-initializes _`extents_`_ with `ext`, and

>   * [9.2]{.pnum} if _`static-padding-stride`_ is equal to `dynamic_extent` direct-non-list-initialized _`stride-1`_ with the least multiple of `pad` greater than or equal to `ext.extent(0)`.

Reviewer: ```suggestion
  * [9.2]{.pnum} if _`static-padding-stride`_ is equal to `dynamic_extent`,
      direct-non-list-initializes _`stride-1`_ with the least multiple of `pad` greater than or equal to `ext.extent(0)`.
```

[19]{.pnum} *Effects:*

  * [19.1]{.pnum} Direct-non-list-initializes _`extents_`_ with `other.extents()`, and

>   * [19.2]{.pnum} is `static-padding-stride` is equal to `dynamic_extent` direct-non-list-initializes _`stride-1`_ with `other.stride(1)`;

Reviewer: ```suggestion
  * [19.2]{.pnum} if `static-padding-stride` is equal to `dynamic_extent`,
      direct-non-list-initializes _`stride-1`_ with `other.stride(1)`.
```

* [20.2]{.pnum} `is_constructible_v<extents_type, typename LayoutLeftPaddedMapping::extents_type>` is `true`.

[21]{.pnum} *Mandates:* `padding_value == dynamic_extent || LayoutLeftPaddedMapping::padding_value == dynamic_extent || padding_value == LayoutLeftPaddedMapping::padding_value` is `true`.

[22]{.pnum} *Preconditions:*

* [22.1]{.pnum} If `extents_type::rank() > 1` is `true`
    and `padding_value` does not equal `dynamic_extent`,
    then `other.stride(1)` equals
    the least multiple of `padding_value` greater than or equal to
    `extents_type::`_`index-cast`_`(other.extent(0))`.

* [22.2]{.pnum} `other.required_span_size()` is representable
    as a value of type `index_type` (*[basic.fundamental]*).

[23]{.pnum} *Effects:*

  * [19.1]{.pnum} Direct-non-list-initializes _`extents_`_ with `other.extents()`, and

>   * [19.2]{.pnum} is `static-padding-stride` is equal to `dynamic_extent` direct-non-list-initializes _`stride-1`_ with `other.stride(1)`;

Reviewer: ```suggestion
  * [19.2]{.pnum} if `static-padding-stride` is equal to `dynamic_extent`,
      direct-non-list-initializes _`stride-1`_ with `other.stride(1)`.
```

[25]{.pnum} *Constraints:*

* [25.1]{.pnum} _`is-layout-right-padded-mapping-of`_`<LayoutRightPaddedMapping>` is `true` or _`is-mapping-of-layout-right`_`<LayoutRightPaddedMapping>` is `true,

* [25.2]{.pnum} `extents_type::rank()` equals zero or one, and

* [25.3]{.pnum} `is_constructible_v<extents_type, typename LayoutRightPaddedMapping::extents_type>` is `true`.

[26]{.pnum} *Precondition:* `other.required_span_size()`
is representable as a value of type `index_type`
(*[basic.fundamental]*).

> [27]{.pnum} *Effects:* direct-non-list-initializes _`extents_`_ with `other.extents()`, and

Reviewer: ```suggestion
[27]{.pnum} *Effects:* direct-non-list-initializes _`extents_`_ with `other.extents()`.
```

either the constraints or the preconditions. <i>-- end note]</i>

```c++
constexpr array<index_type, extents_type::rank()>
  strides() const noexcept;
```

[29]{.pnum} *Effects:* Equivalent to
`return array<index_type, extents_type::rank()>({stride(P_left)...})`.

```c++
constexpr index_type required_span_size() const noexcept;
```

[30]{.pnum} *Effects:* 

>   * returns `0` if the multi-dimensional index space _`extents_`_ is empty, otherwise

Reviewer: ```suggestion
  * returns `0` if the multidimensional index space _`extents_`_ is empty, otherwise
```

>   * returns `*this((`_`extents_`_`(P_left)-1)...) + 1`

Reviewer: Oh wait, what if the rank is >= 2 but one of the extents is zero?

Reviewer: Never mind; that's addressed right above this.
